pkg/ccl/backupccl/backup_metadata.go

```diff
@@ -0,0 +1,1135 @@
+}
+
+// BackupMetadata holds all of the data in BackupManifest except a few repeated
+// fields such as descriptors or spans. BackupMetadata provides iterator methods
+// so that the excluded fields can be accessed in a streaming manner.
+type BackupMetadata struct {
+	BackupManifest
+	store    cloud.ExternalStorage
+	enc      *jobspb.BackupEncryptionOptions
+	filename string
+}
+
+func newBackupMetadata(
+	ctx context.Context,
+	exportStore cloud.ExternalStorage,
+	sstFileName string,
+	encryption *jobspb.BackupEncryptionOptions,
+) (*BackupMetadata, error) {
+	var encOpts *roachpb.FileEncryptionOptions
+	if encryption != nil {
+		key, err := getEncryptionKey(ctx, encryption, exportStore.Settings(), exportStore.ExternalIOConf())
+		if err != nil {
+			return nil, err
+		}
+		encOpts = &roachpb.FileEncryptionOptions{Key: key}
+	}
+
+	iter, err := storageccl.ExternalSSTReader(ctx, exportStore, sstFileName, encOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	var sstManifest BackupManifest
+	iter.SeekGE(storage.MakeMVCCMetadataKey([]byte(sstBackupKey)))
+	ok, err := iter.Valid()
+	if err != nil {
+		return nil, err
+	}
+	if !ok || !iter.UnsafeKey().Key.Equal([]byte(sstBackupKey)) {
+		return nil, errors.Errorf("metadata SST does not contain backup manifest")
+	}
+
+	if err := protoutil.Unmarshal(iter.UnsafeValue(), &sstManifest); err != nil {
+		return nil, err
+	}
+
+	return &BackupMetadata{BackupManifest: sstManifest, store: exportStore, enc: encryption, filename: sstFileName}, nil
+}
+
+// SpanIterator is a simple iterator to iterate over roachpb.Spans.
+type SpanIterator struct {
+	backing bytesIter
+	filter  func(key storage.MVCCKey) bool
+	err     error
+}
+
+// SpanIter creates a new SpanIterator for the backup metadata.
+func (b *BackupMetadata) SpanIter(ctx context.Context) SpanIterator {
+	backing := makeBytesIter(ctx, b.store, b.filename, []byte(sstSpansPrefix), b.enc, true)
+	return SpanIterator{
+		backing: backing,
+	}
+}
+
+// IntroducedSpanIter creates a new IntroducedSpanIterator for the backup metadata.
+func (b *BackupMetadata) IntroducedSpanIter(ctx context.Context) SpanIterator {
+	backing := makeBytesIter(ctx, b.store, b.filename, []byte(sstSpansPrefix), b.enc, false)
+
+	return SpanIterator{
+		backing: backing,
+		filter: func(key storage.MVCCKey) bool {
+			return key.Timestamp == hlc.Timestamp{}
+		},
+	}
+}
+
+// Close closes the iterator.
+func (si *SpanIterator) Close() {
+	si.backing.close()
+}
+
+// Err returns the iterator's error
+func (si *SpanIterator) Err() error {
+	if si.err != nil {
+		return si.err
+	}
+	return si.backing.err()
+}
+
+// Next retrieves the next span in the iterator.
+//
+// Next returns true if next element was successfully unmarshalled into span,
+// and false if there are no more elements or if an error was encountered. When
+// Next returns false, the user should call the Err method to verify the
+// existence of an error.
+func (si *SpanIterator) Next(span *roachpb.Span) bool {
```

Comment:
I don't have any particular objection to these write-to-value-at-pointer style, but I think pebble and pkg/storage _tend_ to do separate-method-to-read-current-value, e.g. `for i.Next() { x := i.Val(); use(x); }`. I don't see a particular difference though.